### ISSUE TITLE
"Versioned" Class Tags

### ISSUE THREAD

**Issue by tolmasky** [facebook/immutable-js#1592]
As some of you may be aware, RunKit built in native support for displaying a number of Immutable classes (see [figure 1](####figure-1) below), and we'd like to increase that support in the future. One thing we might run into though, is that due to the nature of RunKit, it actually requires some internal knowledge of immutable classes. The reason for this is that RunKit does a full object serialization on the backend, and its only the frontend that is recognizing the "shape" of the object, and merely providing an Immutable *view* of the data in question. One cool side effect of this is that you can actually select "properties viewer" in RunKit and see the normal JavaScript representation, which can be helpful if you are debugging immutable itself (see [figure 2](####figure-2) below).

#### Figure 1
<img width="860" alt="screen shot 2018-09-12 at 3 33 21 pm" src="https://user-images.githubusercontent.com/23753/45456877-40ec0680-b6a1-11e8-9319-29bf045054bf.png">

#### Figure 2
<img width="878" alt="screen shot 2018-09-12 at 3 33 26 pm" src="https://user-images.githubusercontent.com/23753/45456938-8c061980-b6a1-11e8-8d49-632e1380ecaf.png">

All this being said, it creates a potential headache in the future (which we may already be running into with 4.0.0?) if the internal representation changes. We can certainly take on the work ourselves of trying to identify the "4" shape vs "3" shape if it happens to be different, but I'd like to propose perhaps modifying the strategy from storing "true" in `"@@__IMMUTABLE_LIST__@@"`, to storing the semver version:

```javascript
{
    "@@__IMMUTABLE_LIST__@@": "4.0.0-rc.9"
}
```

I think this would actually be useful outside of RunKit as well. It is perhaps the case that most of the time two different versions of immutable objects can interact just fine, but this would provide an excellent way of either 1) doing a conversion if two incompatible versions are detected, or 2) being able to easily throw a very friendly error if incompatible versions are detected. I think this is very much in-line with the "spirit" of the class tag: to be able to identify that an object is indeed an immutable list even if it does not `intanceof` to your stored copy of the constructor.

**Comment by conartist6** [comment]
It's pretty cool what you guys are doing.

As to what you're asking, you seem to be saying that while RunKit is happy enough using the sentinel to detect immutable objects, it is actually scraping their contents from the internal representation in order to present the "Immutable view" which shows the stored data.

I guess there's not much point in hypotheticals since your use case is already basically the worst case scenario for Immutable in terms of why any program could want the information you're asking be made part of the public API:  to use it's non-public API as a public API. That isn't really what semver is meant for. On the other hand the semver spec chooses to protect the implementation itself as an API by offering patch versions and forbidding version reuse.

I also understand why you need to do that though, since method calls or even vanilla property accesses have the possibility of changing the program state.

Anyway, it isn't really my call, just offering my thoughts.

**Comment by conartist6** [comment]
Also there really shouldn't be much in the way of incompatible immutable versions because Immutable is designed to use its own most stable public APIs: iteration and those sentinels.

**Comment by tolmasky** [comment]
So, unfortunately I suppose the discussion somewhat forks here between "is this a good idea in general" and "is this necessary for RunKit". Just since I find the discussion itself interesting, if you'll allow me to tackle the first momentarily, it is certainly the case I believe that you should definitely not be mixing v4.x.x immutable data structures with v3.x.x data structures, even if they do only use each other's public API's. I am aware of behavior changes in said public API's, as well as "breaking" bug fixes (which also essentially constitute behavior changes as well). Ultimately, if a set of changes merits a "breaking" semver change, then in theory that should be sufficient to argue that the versions are incompatible (otherwise you could get away with a mere minor change - if for example you are merely *adding* new public methods for example).

I am not aware of the exact discussions around originally adding these sentinel values, but the reasoning I've seen for them in the past is to be able to definitively detect something as Immutable if it, for example, come from a different Realm (the same way `Array.isArray` can identify an Array even in cases where `instanceof Array` is false). The classic case of course are when two IFRAMEs share data, but this also of course comes up with require("vm") contexts in Node.js. In those worlds of course, you can never have two different versions of `Array` though (you wouldn't for example have one IFRAME with a Safari 11.x functioning `Array.reverse` vs. a [Safari 12.x broken `Array.reverse`](https://stackoverflow.com/questions/52390368/array-state-will-be-cached-in-ios-12-safari-is-bug-or-feature/52392901) -- sorry I couldn't resist the fortuitous topical bug!) , they are merely two identical instances. However, it is completely conceivable that two IFRAMEs *could* have two different versions of Immutable.js (due to the... unpredictable... nature of npm installs, or separate teams not communicating, or what have you). As such, it doesn't seem that far fetched to say that libraries should be checking the a sentinel *version*, and turning subtle hard-to-find bugs, into friendly errors that are thrown. Arguably you could store only the semver Major in such a setup if you fully trust that the minor and patch versions truly make no API changes.

Separately, from the perspective of RunKit, we are in kind of a bind. I once again apologize if this was already perfectly clear from my first post, but I thought that it might have perhaps not provided enough context -- specifically, I explained "how" are system works, but necessarily "why" it works this way. In a lot of ways, you can think of RunKit as an even more sophisticated version of the Realm-sharing problem described above, where the originating Realm is a node.js process, and the receiving Realm is the client-side JavaScript process. I think a fair question might be why we don't just figure things out "on the server" where we have complete information. The answer to this is many-fold: 1) the server may "happen" to be running a user's immutable library, but to do what we're describing would many we'd have to put a lot of immutable-specific intelligence into the backend (which while not elegant), more importantly slows down the "just run the code" part, vs the front-end which can take as long as it wants to then figure things out and render. 2) Separately, one really cool side-effect is that all the old documents that were written before we had Immutable support magically learned how to display this Immutable data as well -- since the outputs of Notebooks never need to be "upgraded" or changed - they are very much just a serialization of the memory generated. 3) The ultimate end goal for us is to eventually allow any library to write custom viewers, and as such we want a system where viewer-writers are "working in one place" -- the client. Not doing half the work on the server to just then turn around and do the other half on the client. We have every intention to someday make this Immutable viewer open source (the only reason we haven't already is that it would be open source for an API that is so far not accessible to users). And, finally, 4) even if we could do stuff on the server we'd run into similar issues since RunKit is happy allowing you import multiple versions of the same library. Again, it may seem silly to do this at the top level ([example](https://runkit.com/tolmasky/json-diff-example)), but it often arrises naturally from other libraries internally requiring differing (non-dedupable) libraries. So, even in a server world we'd still potentially run into this problem.

All this being said, our position is very much that this is "our problem", and completely understand if you do not like the idea for the library. It is already the case that our "shape matching" goes beyond checking for the sentinel, if for no other reason than we have to make sure we don't try to display a false positive and render nonsense! So in a world where this was not deemed like worthwhile addition, we would certainly take the time to look at the way the internal representation changed on every version and make sure our renderers could account for it. However, it would make our lives much easier if the (I believe) small change of changing the value of the sentinel value from a bool to a version string were made.

Anyways, I know this was a wall of text and I certainly appreciate you listening to all this either way, the library is great and we use it for absolutely everything we do at RunKit!

**Comment by conartist6** [comment]
So the more I think about what the architecture of your system must be like, the more I think you should just have the backend interact with Immutable and send along the resultant data.

You're presumably using Object.getOwnPropertyDescriptor and Object.getPrototypeOf when you encounter object-like things in order to avoid triggering side effects. I don't even want to think about the special cases you'd encounter for arrays. But Map and Set you would have to recognize as data storage types which are safe to iterate over, at very least using something like `Map.prototype[Symbol.iterator].call(someMap)`. The support that Immutable gives you is that it tags its structures with sentinels and hides its constructors. This is a relative guarantee that `someImmutableMap[Symbol.iterator]()` is going to be pure. You'd end up having to send two representations of your object from your backend presumably -- the object structure one and the iterable data one -- but in return you get what you came here for, which is that your approach will be supported by Immutable, as well as by other data structure libraries (like [Sequins](https://github.com/conartist6/sequins) which I am working on), and could even share some code with es6 Map and Set serialization.

**Comment by conartist6** [comment]
I see, so RunKit is working on the C++ side, not in JS-land?

**Comment by tolmasky** [comment]
It's a pretty complicated mixture. To guarantee true purity from displaying outputs, we for example will fork the memory of the process and operate on the fork (a good example of this is capturing stack traces of Error objects -- it is unfortunately the case that calling `.stack` is a *mutating* operation in v8, so attempting to serialize it would break later "user-level" callers like source-mapping libraries). Similarly, getting the internal state of a Promise without attaching a `.catch` (which thus would affect whether you get an unhandledPromiseRejection) also requires C++.  But yes, we do actually end up having to use C++ for just about every object for certain properties.

All this being said, I think a more "conceptual" reason is that our serializer would become a bit of a mess if we started trying to do different serialization for every library under the sun. RunKit's module-fs has every possible installation of every version of every package on npm. Client-side renderers is simply the most reasonable way to support everything without slowing down the actual *operation* of the program.

Additionally, it would mean that our serializer changes from a series of well trusted internal methods, to one which fundamentally hands of the process to arbitrary code. The display of a value could behave in unpredictable ways (never terminating, crashing, changing the state of the very objects its observing, etc. etc.). But let's say temporarily that we mark Immutable.js's iteration methods as "blessed" functions on the same order as the JavaScript built-ins, since they are in fact theoretically "immutable". As you already started to account for in your suggestions, we still couldn't just "trust" `theMap[Symbol.iterator]`, since the user could have changed the contents of `Symbol.iterator`. But we actually also can't trust `Map.prototype[Symbol.iterator]` *at this point* because the user could have changed that too. So what we'd *really* want, is a copy of the *original* `Map.prototype[Symbol.iterator]`, which you can only get safely *before* any of the user's code has run. This immediately means that you are changing the order in which `requires` take place in the code (sidestepping the complexity of determining whether the code uses immutable in the first place to even know to do this ahead of time). And again, this problem is further exacerbated by the fact you may need to grab *multiple* versions of all these methods since the user's code may (either explicitly or inadvertently) grab multiple version of Immutable (perhaps there is a class that exists in one version but not another). Just to give you a sense of the complexity we already deal with just with built-in methods, it's not enough to store `Array.prototype.reduce` and do `ArrayReduce.call(theArray)` for example, because the user could have changed `Function.prototype.call`! So the way to safely call reduce is:

```javascript
// At the top before any other code (and thus before bind could be changed)
var Call = (Function.prototype.call).bind(Function.prototype.call);
var ArrayReduce = Array.prototype.reduce;

// ... later

Call(ArrayReduce, theArray);
```

So, its already quite difficult to maintain the constraints we promise with just the built-in methods, let alone accounting for an unbounded number of Immutable methods, which would of course then not scale very well to all the other libraries on RunKit. It would very much mean RunKit intrinsically supports Immutable in a super special way.  And, again, it means that the viewer is fundamentally tied to the time the document was run, vs. the IMO more elegant world where the viewer can be updated or added after the fact.

Hopefully this gives some sense for what's going on, and why the client-side model is where we landed. Want to re-iterate though that we do not want to force any decision on you all just due to the very particular set of problem we are trying to solve.

**Comment by conartist6** [comment]
We're not talking about "an unbounded number of immutable methods". We're talking about one method, `Symbol.iterator`. You shouldn't even worry about people overriding `Symbol.iterator`, since it's meant to be overridden. I guess you would want to make sure that you get the original value of the iterator symbol itself. Then when someone tries to `console.log` an Immutable object (or any other non-primitive iterable object!) you just to fork the process and run the method defined by the iterator symbol. It seems to me like you'd end up with robust flexible code. Then all you'd need to do for immutable would be to look at the sentinel values to decide whether to format entries, for example turning`["key", "value"]` into `key => "value"`. You'd also support my library, Sequins, and any number of other tools which use iterables, which should eventually be lots and lots of them! [Here](https://runkit.com/conartist6/5ba7b312f20fef0012598374) is my test on Sequins, by the way.

However: if markers are indeed the solution, I'd make them RunKit specific, something like `_runkit_internal_structure_ver`. That way Immutable would be accepting your use case, but hopefully not open the door for everyone to do similar things. Even better than an obscure name would be an npm module which would be a dependency of both RunKit and immutable which exported a symbol which would be used to access the internal structure version number. I assume that RunKit always executes in an environment which has native support for Symbols.

**Comment by conartist6** [comment]
I should say also that I plan to reuse much of the format that Immutable's documentation uses for Sequins since the two projects share [most of] the same API. That would include RunKit examples, so to make it extra clear I have personal interest in a solution that isn't just a one-off for Immutable.

**Comment by conartist6** [comment]
I'm still thinking about this. I realize something that should have been obvious to me: consuming iterables is not generally a safe strategy since they may well be infinite. `toJSON` would work though, and implies that the object has an obvious JSON representation. It doesn't have some of the cleaner properties you describe in terms of having complete backwards compatibility by virtue of only sending memory serialization to the frontend, but you're going to have a hard time convincing me that backwards compatibility is worth more than forwards compatibility.

**Comment by conartist6** [comment]
Actually `toJSON` has the same problem in Immutable: `Range().toJSON()` is going to break. Have you considered the possibility of something like a `__toRunkit__()` method?

**Comment by tolmasky** [comment]
Our constraints are that we cannot modify the library. I think I've outline the reasons above, but to re-iterate, unless you are proposing we add `__toRunKit__` to immutable.js proper (which I am sure you are not), RunKit needs to simulate the normal running environment and thus cannot go adding magic methods everywhere that then won't appear when you try to run the same script locally. blah.__toRunKit__ can't be truthy in RunKit and falsey in node. Separately, again, all this is already predicated on the idea that we would do a bunch of library mutation if we detect you are specifically using immutable.js, which is just one of the 500,000 (x however many versions) of npm packages supported on RunKit. I am totally OK with the answer being "we don't want to do this". We'll just do shape detection on the client, or get rid of the feature.

**Comment by conartist6** [comment]
Why not add it to Immutable.js proper?

**Comment by conartist6** [comment]
I'm not against a pattern of supporting Runkit directly. I just think that any such solution would best be part of the "public API", at least in the sense that semver dictates that semantic versioning applies specifically to the published public API of a library. I also think that a solution should not be specific to Immutable.js. If `__toRunKit__` being part of Immutable would work for you I'll write the PR myself and advocate for it.

**Comment by tolmasky** [comment]
Let's use `Range()` as an example. Right now this what RunKit shows:

<img width="376" alt="screen shot 2018-10-20 at 9 56 18 am" src="https://user-images.githubusercontent.com/23753/47258315-6c0b0800-d44e-11e8-9139-87e2ecb59ef6.png">

The problem of "serializing an infinite Range" is *only* difficult if we insist on relying on iteration methods (which then leads to needing custom stuff). RunKit has all the data it needs right there. If `Range` had a `{ "@@__IMMUTABLE_RANGE__@@": true }` marker, the client can trivially see the start, step, and size, and display something like this:

<img width="370" alt="screen shot 2018-10-20 at 10 00 27 am" src="https://user-images.githubusercontent.com/23753/47258360-feaba700-d44e-11e8-9f58-7e568e4c9331.png">

So here, I'm just requesting it have the same marker as `List` currently has. If we then make it store "3.5.8" instead of "true", then I can account for `_step` changing to `_byStep` in a later version.

Sure, we can invert this and have `__toRunKit__` return a "standardized" serialization of, say, `{ type: "Range", size: "infinity", "start": 0 }`, but then that 1) requires super specific RunKit code in immutable.js (which sure, I'm flattered if you would include), but 2) requires us to *run code during output* which we do not want to do for the reasons outline above:

1. We don't want to have the serialization process hand-off to user-defined functions that are fundamentally black boxes (our current serialization relies solely on v8 defined `getObjectDecscriptor, etc. which have known perf characteristics/known to terminate/known to not throw/etc.)
2. We don't want to have to do "is this the `__toRunKit__` I expect checks?" of storing `Range.prototype.__toRunKit__` before any code is run to avoid users that do `myRange.__toRunKit__ = hahaFunc`.

**Comment by conartist6** [comment]
Here's the problem I see though. Say the sentinels with versions are added. One of two things happens.

1. You publish an open source serialization-from-memory library for Immutable. Other libraries can publish their own external deserialization libraries to get RunKit support. Now Immutable's internal structure essentially part of it's public API. People will find uses for the serialization library that aren't Runkit, and ever library which opts to create a public runkit serializer will have the same problem. Changes will be added in a patch version that affect the memory structure, most people will have installed the package with a caret-based range and will automatically end up with the new version, and then silently functionality will break.
2. The serialization-from-memory code is a private part of Runkit, and you're again left supporting Immutable specifically instead of having a generic solution which allows the community to embrace Runkit instead of Runkit embracing the community.

I also don't quite see why you think you need to worry about what the value of the `__toRunKit__` function would be. My thought is: let's say your code itself needs to filter arrays. You need to protect yourself from the possibility that `Array.prototype.filter` has been modified by the user. Perhaps in a flash of brilliance they exchanged the method for one that, say, always filters out null, and now Runkit itself is completely busted. That is a failure of defensive coding for Runkit's javascript. Calling a method which is meant to have the sole function of defining the behavior of the object it is called on does not have this problem.

Reason 1 is certainly valid though. You'd have to be defensive there still and fall back to the regular memory serialization if the `__toRunKit__` method threw an error or didn't complete within, say, 100ms.

**Comment by tolmasky** [comment]
There exists no "generic" solution. The options are specialized RunKit code for Immutable on the server (which hopefull I will show you is much more cumbersome), or custom Immutable code on the client. `__toRunKit__` is not usable in the general case. Let's look at an example:

Let's say your notebook starts off like this:

<img width="750" alt="screen shot 2018-10-20 at 10 17 27 am" src="https://user-images.githubusercontent.com/23753/47258544-7d094880-d451-11e8-8a42-783c07f4fd07.png">

So far so good. Nothing crazy. Now let's say you append a new cell that references `immutable`:

<img width="757" alt="screen shot 2018-10-20 at 10 17 36 am" src="https://user-images.githubusercontent.com/23753/47258547-8e525500-d451-11e8-9ee8-9d7a7327636b.png">

As you can see, the user inadvertently added code that would lead to an infinite loop in log serialization. Again, the particularities of *how* the user is breaking serialization don't necessarily matter. It could have just included a logic error that results in an exception, or worse, the process to exit if they call into a node binary module which hard crashes.
There is no reasonable defense for this aside from treating trivial logging as potentially catastrophic behavior that requires essentially restarting the process for, which still is a non-answer since you're still broken at that point,
just responsive -- you'll never log past it. This scenario simply doesn't exist with client-side display strategies.

But, let's see how far we can get. Let's say we pretend that we can trust "the real `__toRunKit`" if its the one actually written by the Immutable.js authors because we trust them. So we want to only run the *original* `Range.prototype.__toRunKit__`. We'd have to, at some point before `__toRunKit__` is changed and thus lost forever, store both `Range` and `__toRunKit__`:

```javascript
const ImmutableRange = I.Range;
const ImmutableRangeToRunKit = ImmutableRange.prototype.Range;
```

We need to store `I.Range` because its the only way we'd know to even run `__toRunKit__` now, since we're no longer just calling it whenever we see it:

```javascript
// ...
if (value instanceof ImmutableRange)
    return Call(value, ImmutableRangeToRunKit)
// ...
```

  1. So, right off the bat: we're now going to be doing several if/else branches of instanceof checks in our serialization which is incredibly bad for performance. Not to mention, if the user requires multiple versions of immutable, we need to now start doing instanceofs for each version, since we have to run the *right* `__toRunKit__`:

  ```javascript
  // ...
  if (value instanceof ImmutableRange_3_5_7)
      return Call(value, ImmutableRangeToRunKit_3_5_7)
  // ...
  ```

2. When and where are we supposed to store this information? We can't do it any point after line 4, because it'll be to late since `__toRunKit__` is gone. We can't do it before line 2, because then the line 2's log will be incorrect (it will log 1 instead of 0). I suppose the "correct" place is theoretically between line 2 and 4 (*in this one particular case where requiring doesn't have other unexpected side effects), but that's obviously difficult to semantically generate. So I suppose we could now do some sort of require-hook magic, that specifically knows about "special" packages like immutable, and it also knows the special exports that themselves contain `__toRunKit__`:

```javascript
function customRequire(path)
{
   if (path === "immutable")
  {
     const exports = originalRequire(path);
     classToToRunKitMap.set(exports.Range, exports.Range.prototype.__toRunKit__);
     classToToRunKitMap.set(exports.List, exports.List.prototype.__toRunKit__);
     //...
  }
}
```

OK, well now we've at least found a generic solution to the versioning problem referenced in (1) I suppose, since our code now looks like:

  ```javascript
  // ...
  for (const constructor in classToToRunKitMap.keys())
     if (value instanceof constructor)
        return Call(value, classToToRunKitMap.get(value));
  // ...
  ```

But, now we're right back where we started. This only works because we *knew* that "immutable" was a blessed package, and not only that, we knew exactly where all the `__toRunKit__`'s in Immutable were ahead of time. This is not generic in any real way, we'd have to continue white-listing packages (which we vetted the `__toRunKit__`'s for!), and additionally know whihc specific exports had `__toRunKit__`s (exports.List DOES, exports.Collection DOESN'T). There is *no way* for me to handle `__toRunKit__` from arbitrary objects, and if your package generates classes dynamically (class expression), then I'll never have a good hook into that either. I'm really just writing custom server code for Immutable.

**Comment by conartist6** [comment]
Please bear with me, I don't have access to the Runkit sources or other materials about its workings which would make it easier for me to experiment and understand its architecture, problems, and solutions.

As I stated, any solutions which involve "blessed" versions of `toRunKit` extracted off of prototypes are useless from both of our perspectives. My suggestion only works if you can execute `__toRunKit__` without trusting it. Your point, and I'll have to take your word for it, is that that is not possible, or perhaps not possible to do without critically compromising performance.

Since it is proving not to be productive for me to speculate on what RunKit could do, perhaps you could speculate with me as to how to avoid the memory structure of any supported library essentially becoming its public API. Are we on the same page regarding this? Do you agree that it is a problem?

**Comment by tolmasky** [comment]
Sure, I guess just so we can be on the same page in totality before moving on to the memory structure stuff, want to summarize the issues with calling arbitrary `__toRunKit__`s as it is not just performance. There is also crash potential and infinite loop potential. Crashes speak for themselves, there's not much we can do in that case. With regard to infinite loops, since they could take place synchronously, it is not trivial to just place timeouts. We're now either talking doing things on threads, or having external monitors that would still only at best be able to kill the process, there's no way for me to "halt" a `__toRunKit__` that has `for(){}` in it without also halting the process itself, short of fairly complicated solutions of throwing the serialization into a thread or worker which don't play well with arbitrary objects (especially binary modules like canvas).

Leaving that aside, I guess I'm not really sure how using a version number exacerbates this "public API" problem in any way, and in fact would ironically be the only defense *against* this scenario.

Immutable objects already have a publicly accessible sentinel value. I'm already using it to detect these objects. The problem already exists. But this goes beyond the sentinel value. Every member variable in Immutable objects share this same problem. Unless you start using `WeakMap`s for all private variables in your JavaScript objects, every property on every object already behaves as "public API" as you are describing. People may very well be relying on `_step` in their code. This is just the way JavaScript works, it (currently) has no private variables, so your best bet is hint to people not to use things by using underscores or Symbols, but this is no guarantee at all. Again, if we really care about *this* problem, we should be talking about removing all underscore properties and replacing them with WeakMap to truly make the objects black boxes.

The interesting thing about taking *the existing sentinel value* and merely making it a string with a version number, is that you can *at least now know* if your private fiddling will probably break. If I have code that I wrote when I was using in version "3.5.7" that uses `_step`, then by checking if `"@@_IMMUTABLE_RANGE_@@"` equals "3.5.7" I can at least know that I'm working against the same version that I originally inspected, or bail if its a different version. Version sentinels are actually the safest of the already unsafe "fake private" properties, and pass this enhanced safety to everything else. In RunKit, if we wrote our code for custom viewers against "3.5.7", beyond the sanity shape check, we could additionally simply degrade to non-custom-viewers if the version string is different. At worst, you get the non-custom view. We then, at our leisure after the fact, can write the "4.0.0" viewer, which would *retroactively* work on all documents that used the new version of Immutable before we had time to update our code. Neither the Immutable authors nor the RunKit authors would be blocked on updating viewer or serialization code to push new versions -- they are decoupled.

**Comment by conartist6** [comment]
It certainly is the case that people can access and use `_step` if they want to. I don't think that's a problem due to the contract created by the semver spec. But it could as easily be public API changes that a library wants to use a semver check on in order to support multiple versions, so I guess that's OK.

So let's say that we do flags. My question is: where does the serialization code live? As a library author can I contribute a serialization of my library to Runkit?

**Comment by tolmasky** [comment]
I may not understand (apologies if so), but I think the serialization is now unrelated to this problem -- or rather, there is no specific serialization requirement. For the "general" purposes of this discussion, it could very well be JSON. RunKit happens to use something more sophisticated since JSON doesn't support NaN, -0, Infinity, circular references, etc. If and when we open source that serialization library, it will much like `JSON.stringify`, be orthogonal to Immutable -- all it does is recursively traverse the own properties of an object and its prototype, and so forth up to the last parent, storing all said information. So for example, `serialize(Object.create(null, { a: { value: 5 } }))` produces 

```JSON
{
  "type": "object",
  "className": "Object",
  "properties": {
    "@a": {
      "key": "a",
      "flags": 0,
      "value": {
        "type": "number",
        "value": 5
      }
    }
  },
  "prototypeOf": {
    "type": "object",
    "value": null
  }
}
```

So, like `JSON`, the location of the library is neither tied nor related to immutable. It is only the UI library, which operates on the reconstructed object (which can be derived from the above serialization), that checks for the presence of the sentinel.

**Comment by conartist6** [comment]
Maybe I used the wrong terms. Immutable gets special treatment by Runkit. If other libraries want a nice display format for a complicated internal structure, can they write the code which does it?

**Comment by tolmasky** [comment]
Yes, but it will not be through any sort of injection into the serialization, but rather through our **Viewer** API. Required packages get a shot at telling us whether they "want" to display any particular data *on the client*. Allow me to shed some background to explain. Any particular piece of data may have multiple ways to view it. For example, the array `[37.7576793,-122.5076403]` has three qualifying viewers: Map Coordinates, Array, and Chart:

<img width="781" alt="screen shot 2018-10-20 at 3 10 32 pm" src="https://user-images.githubusercontent.com/23753/47260976-590e2d00-d47a-11e8-979e-5928cd734128.png">

As you can see, by default we display it as **Map Coordinates** because it looks an awful lot like latitude and longitude. You can alternatively just look at it as a normal Array object or as a Chart. The Chart is chosen last because it thinks you probably don't want to see a bar graph of two items. Notice however that `[1, 0]` does not choose **Map Coordinates** by default, but rather the standard object viewer:

<img width="758" alt="screen shot 2018-10-20 at 3 12 14 pm" src="https://user-images.githubusercontent.com/23753/47260983-92df3380-d47a-11e8-9f31-ff6240ac5fc0.png">

That's because, while technically [1,0] is a valid geographic coordinate, it's not really likely thats what you wanted to see (most geographic coordinates have more precision, etc.). Incidentally, the object `{ latitude: 37.7576793, longitude: -122.5076403 }` will also display as a geographic coordinate by default, but of course not have the Chart option:

<img width="759" alt="screen shot 2018-10-20 at 3 38 39 pm" src="https://user-images.githubusercontent.com/23753/47261163-69280b80-d47e-11e8-9ddc-e964bcece089.png">

Again, here we are reasonably certain that you're dealing with a Map Coordinate, but you can look at it like a normal object too.

## Custom Viewers

RunKit has a wide variety of users, from data analytics where you are dealing with raw data where this kind of inference can be very helpful, to general node debugging where the structure of the object is more concrete. Our Viewer API is flexible enough to handle both, and ends up showing that both can actually be useful in both domains too. The basic way all this works on the client is that you are handed each object that will be displayed, and you have an opportunity to tell us whether you'd like a shot at showing it, and how confident you are that you are the best choice. You can imagine this confidence value as being anything between 0 (can't show it at all) to 1 (you are absolutely certain you are THE viewer), but its most helpful to think in terms of CANT (0), CAN (0.5 - you could show it if you wanted to, but there's nothing that screams you should be the first choice), and WANT (1 - again, you are the best choice). Once you are selected as the viewer, we hand the object to you to display however you see fit.

In the case of Map Coordinates, a very high confidence value is returned for objects that have just a latitude and longitude property with number values within the allowable geo ranges. Slightly lower for arrays with high precision within those ranges, and very low, but not zero, for arrays that could be displayed that way but don't really seem like it.

In the case of something like Immutable, we are really only concerned with 0 and 1: we're either going to be more or less certain we're a `List`, or we're not going to be able to do anything with it at all. The way we determine this is by checking that said object has the proper sentinel value (`"@@_IMMUTABLE_LIST_@@"`), and then validate the internal structure (notice that RunKit won't try to show just any object that has an `"@@_IMMUTABLE_LIST_@@"` key as a `List`), and then if it all checks out, chances are this thing was a `List`, and you return 100% confidence basically (And if it wasn't and the user somehow magically assembeled an object that has a completely compatible structure, then that doesn't seem so bad either ;) ).
The API actually lets you use JSON schemas for this, so it can be rather trivial and declarative to describe the objects you want to handle. All this would end up looking something like:

```javascript
myListViewer =
{
    test: someValue => /* return 1 if it looks like a List */,
    display: someValue => /* show the List */
}
```

This gives package authors not only the ability to detect, but also the ability to actually write their custom viewers. All this viewer testing and displaying takes place in iframes, so you *can* time them out, let them fail, etc. without affecting your main site.

Currently we are taking on the mantle of displaying the Immutable values as this is not public yet (and won't be for a while - lot of details not present here), so it seems like we're using your private data. But once this API is public, it would be the same package authors, viewing their own private internal state on the client. It is **these viewers** that we would contribute to your project (which you could choose to incorporate or not). So its not really that we're trying to publicize your internal sctructure, the idea is that it is the same authors viewing their own serialized internal structure over the wire. Here using version strings would be useful to you. If the user requires both immutable@3.5.7 and immutable@4.0.0, there will of course be *two* potential viewers, and each will test each value that gets shown. The 3.5.7 viewer should show 3.5.7 values, and the 4.0.0 viewer should show 4.0.0 values. It makes it dramatically easier to apply the right analysis if the sentinel value tells you which version you are.

**Comment by conartist6** [comment]
I still really have no idea how to approach this. My thoughts:

- Immutable could in theory separate out its trie implementation into another library and require that library using a semver range. Arguably this would be a step down the path of best practices for Immutable, but it could again mean that a single version of immutable could have more than one possible internal memory structure.
- The viewers would absolutely need to live in some repository you set up. This is largely because of the rather exotic build tooling that would be needed to ensure they are tested correctly, and because they will need to be their own packages, otherwise how could you know which versions of a library contained the active viewer implementations?
- You also describe separate viewers for separate major Immutable versions, and viewers running in iframes on the frontend each processing potential structural matches and outputting a score. I noticed that the structural matches can even be nested. The perf implications of this seem like they would ultimately be catastrophic after the API for viewers becomes public and a significant number of libraries and library versions are involved. There'd just be too many viewers to load! At very least maintaining runtime perf would seem to require all libraries to version flag in the way that you're requesting Immutable does so that they could at least render a rapid 0/1 decision on applicability. It sounds like a pattern with the potential to do damage to semver itself.

**Comment by tolmasky** [comment]
So we keep alternating between highly conceptual concerns, which we seem close to resolving at points (such as the discussion of "public APIs"), to reverting to questions about RunKit display specifics. I understand this is probably my fault for just jumping to our own specific use case at the beginning of this bug, but I think it makes for a complicated discussion (I think this bug is now completely inaccessible to anyone else that would ever want to look at it).

Do you just happen to live in San Francisco? Can we just meet in person? At this point, we're increasingly getting into just absurdly specific implementation details of parts of the product that are further and further removed from the request of this bug. If we're going to do that I'd rather just show you the code then continue to discuss it in prose and screenshots. I want to think there is just some big miscommunication that must be happening here, because I don't know how we go from "hey, if this existing sentinel value could change in this incredibly minor way, it both has some interesting utility for your own internals as well as making our lives way easier -- but don't worry, if you don't want to do that we can figure it out on our own too" to "we need to write a separate trie library". Maybe if we could just sit down together and I just show you these viewers and we interact with them live we could get to the bottom of this.

If that's not practical though, then could we instead just choose to keep the discussion limited to pure library design concerns? For a change like this, I really don't think it needs to be done just "for RunKit", and if I am unable to make a convincing argument about the the utility of this change outside of RunKit, I think it would be a fair reason to have the issue closed, because I really do believe in the importance of this change regardless of the peculiarities of our viewer system. To reiterate an earlier point, I believe that sentinel version strings are a reasonable enhancement of the existing sentinel system regardless of the existence of RunKit. If through some combination of sub-dependencies I end up with two copies of different versions of immutable.js in my project, I absolutely want myList_2.concat(myList_3) to at least warn that there's something probably misconfigured in my program to end up this way. This is for the record not an unlikely scenario, I end up with multiple *copies* (although often the same version) of immutable.js in my project often (and it is one of the reason the sentinel values exist in the first place - to make it so that two lists from non-object-identical List prototypes to be able to cleanly interact). Making the sentinel system carry more information would absolutely make this system more robust. Again this can happen "accidentally" through edge cases in deduping, or "legitimately" through interactions between different realms (such as iframes or v8 vm-contexts). So if we believe in the importance of sentinel values at all, I think we should believe in this change.

**Comment by conartist6** [comment]
Yeah the topic has indeed become extremely and perhaps uselessly broad. I'll just say that I think `myList_2.concat(myList_3)` is just fine, and perhaps that is indeed the root of our disagreement. Anyway, I am in SF and would be happy to meet and discuss in person. I'll send you an email.

**Comment by conartist6** [comment]
So I consider Immutable to have a core API which transcends major versions. The API for es6 `Map` and `Set` is set in stone at this point - namely the `Symbol.iterator`, `keys`, `values` and `entries` methods. I think those methods in Immutable (plus the List versions) can also be considered stable across major version boundaries. The only other part of Immutable needed for it to interact with itself is the sentinel code. Non-backwards-compatible changes to sentinels would be catastrophic for Immutable, and the surface area of that API is so small that I don't worry much about it.

My view then is that if you were not responsible for constructing an Immutable object (and can't be sure which version of the library was) that is the API to which you should limit yourself. I consider `myList_3.concat(myList_2)` to be safe because the API mentioned above is all `concat` needs (by design) to work with the object passed it.

**Comment by tolmasky** [comment]
I think the fact that we are discussing one very such change in [this bug](https://github.com/facebook/immutable-js/issues/1585), where in a new major version the behavior of `.size` will change from lazy to not lazy makes me skeptical of this idea of "major version transcending APIs". `size` seems pretty darned low level of a feature, and even it wasn't immune. Constraints change, lessons are learned, and we can't really know what we'll want in the future.

But even if we grant this notion of APIs so perfectly conceived at the beginning their intended behavior will never possibly change, this still **also** assumes that those APIs will *never have bugs*. APIs break both on purpose and accidentally. Major version changes represent "this code once did this (whether intended or not), but now does this, please change your code accordingly". If we some day discover a bug in `concat` that would significantly alter the behavior of programs by shipping a minor or patch version, then it should of course be shipped as a major upgrade, otherwise entire programs would break when consuming what they believed to be non-breaking updates.

**Comment by conartist6** [comment]
I mean Sequence.size was never safe at all, regardless of version. I don't see why that would be a factor. And concat is not part of the core API that I trust regardless of version. Of course there may be a bug in it or its behavior may change, that's why I wouldn't trust it unless I know which version of it I'd be using.

**Comment by conartist6** [comment]
Let's say Array didn't have the suite of prototype methods it does, and you're dealing with an array created by someone who polyfilled in `Array.prototype.concat`. There is a bug in their implementation, and they write `[1, 2].concat([3, 4])` but the end (bugged) result is `[1, 2, 4]`. Now they pass the array to you. You don't trust the concat method (you're defensive and it could have, and was, changed), but you can't claim that you should know to distrust the content of the array. It's as good an array as any other.

**Comment by conartist6** [comment]
My position is simple. If you don't know who required it and what version of the library they were using, don't trust more than the core API I discussed. If you did require it yourself, refer to the documentation for how to use it. Sorry if I misrepresented or overcomplicated anything.

I'm saying that if I did this:
```js
const {List} = require('immutable');
module.exports = function(unknownVersionList) {
  return new List().concat(unknownVersionList);
}
```
I consider it safe because Immtuable's own implementation of concat only interacts with `unknownVersionList` using the core API which I consider to be safe.

**Comment by conartist6** [comment]
This is great: it's concrete, and it's the right conversation to be having. I didn't know about this change either. Your example is also great at capturing a situation where a major version difference will cause a change in behavior that is very likely to be unexpected.

How would you use versioned tags to warn the programmer?

**Comment by conartist6** [comment]
The versioned tags don't seem to help with defensive programming here because it's too late to implement a forwards compatibility approach to the issue and this change is easy to detect anyway. To cover the backwards compatibility aspect just do the isCollection and Record.isRecord checks to find out what you're dealing with - at which point it may be suitable to warn that since you are concatenating a pre-v4 record the results will be different than you might expect. This seems like a good warning to me -- it tells the coder about something specific which we are sure has happened. The specific version is less important, the programmer will surely figure it out if they need to.

What versioned tags would allow is the addition of a feature in v4 that, whenever immutable detected mismatched major versions being used with each other, would log to the console (only once, though that would be difficult to accomplish reliably) a message stating the two involved versions of Immutable and linking to a page which described known cross-version incompatibilities and gotchas, a page which of course could be kept up to date independent of particular releases.

It still seems like there would be a lot of complexity in the implementation -- there could be more than two versions of Immutable. Versions prior to tagging might be v3 or pre-release v4. And if things aren't broken (yet), I feel like many people will just ignore the message as more console spam and then forget it's there by the time they have fallen victim to a real issue. But it would be helpful to some people, and it isn't as problematic to me as actually triggering different behavior based on versions.
